### PR TITLE
docs(nnx): add native LoRA Gemma4 and Qwen3 notebooks and update tutorial guides

### PR DESCRIPTION
**Pull request by RexBearIU**
# Description

Adds native Gemma4 and Qwen3 (0.6B) LoRA and QLoRA tutorial notebooks for Flax NNX, along with accompanying post-training setup documentation. Also includes a follow-up fix for `lora_llama3_demo.ipynb` from PR #4708.

These tutorials cover both standard PEFT workflows:
1. Pathway A: Native LoRA Supervised Fine-Tuning (SFT) starting from converted Hugging Face checkpoints.
2. Pathway B: Native LoRA Pre-training & Checkpoint Restore with QLoRA.

### Summary of Changes:
- `src/maxtext/examples/gemma4_native_lora_demo.ipynb`: Gemma4 Native LoRA and QLoRA demo notebook.
- `src/maxtext/examples/qwen3_native_lora_demo.ipynb`: Qwen3 Native LoRA and QLoRA demo notebook.
- `src/maxtext/examples/lora_llama3_demo.ipynb`: Follow-up fix from PR #4708 to set `MODEL_NAME = "llama3.1-8b-Instruct"`, ensuring checkpoint conversion downloads the Instruct-aligned weights (`meta-llama/Llama-3.1-8B-Instruct`).
- `docs/tutorials/posttraining/native_lora.md`: Guide for native Flax NNX LoRA.
- `docs/guides/run_python_notebook.md` & `docs/tutorials/post_training_index.md`: Updated documentation indices.

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

All tutorial notebooks (`qwen3_native_lora_demo.ipynb`, `gemma4_native_lora_demo.ipynb`, `lora_llama3_demo.ipynb`) have been extensively verified and executed cell-by-cell under standard TPU VM environments using `papermill` with zero errors (100% clean exit code 0).

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by RexBearIU** [comment]
Rebased branch onto main and fixed Sphinx documentation build warnings regarding notebook links.

File: src/maxtext/examples/gemma4_native_lora_demo.ipynb [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+   ],
+   "metadata": {
+    "id": "9904d180"
+   }
+  },
+  {
+   "cell_type": "code",
+   "source": [
+    "subprocess.run([\n",
+    "    sys.executable, \"src/maxtext/trainers/pre_train/train.py\", \"src/maxtext/configs/base.yml\",\n",
```
Comment (shralex) [inline src/maxtext/examples/gemma4_native_lora_demo.ipynb]: please use -m notation here and elsewhere:

sys.executable, "-m", "maxtext.trainers.pre_train.train", ...

File: src/maxtext/examples/qwen3_native_lora_demo.ipynb [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "MODEL_NAME = \"qwen3-0.6b\"\n",
+    "BASE_OUTPUT_DIRECTORY = \"/tmp/qwen3_sft_output\"\n",
+    "HF_CHECKPOINT_PATH = f\"{BASE_OUTPUT_DIRECTORY}/qwen3_06_checkpoint\"\n",
+    "\n",
+    "if not os.path.exists(HF_CHECKPOINT_PATH):\n",
```
Comment (shralex) [inline src/maxtext/examples/qwen3_native_lora_demo.ipynb]: consider checking for the inner checkpoint path:

ORBAX_ITEMS_PATH = f"{HF_CHECKPOINT_PATH}/0/items"
if not os.path.exists(ORBAX_ITEMS_PATH):

File: src/maxtext/examples/gemma4_native_lora_demo.ipynb [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+   }
+  },
+  {
+   "cell_type": "code",
+   "source": [
+    "MODEL_NAME = \"gemma4-e2b\"\n",
+    "BASE_OUTPUT_DIRECTORY = \"/tmp/gemma4_sft_output\"\n",
+    "HF_CHECKPOINT_PATH = f\"{BASE_OUTPUT_DIRECTORY}/gemma4_e2b_checkpoint\"\n",
+    "\n",
+    "if not os.path.exists(HF_CHECKPOINT_PATH):\n",
```
Comment (shralex) [inline src/maxtext/examples/gemma4_native_lora_demo.ipynb]: here too

**Review by shralex — APPROVED** [review]

File: src/maxtext/examples/gemma4_native_lora_demo.ipynb (hunk 1 above)
Comment (RexBearIU) [reply src/maxtext/examples/gemma4_native_lora_demo.ipynb]: Thanks for catching this. Switched all python script executions over to the module format in both demo notebooks.

File: src/maxtext/examples/qwen3_native_lora_demo.ipynb (hunk 2 above)
Comment (RexBearIU) [reply src/maxtext/examples/qwen3_native_lora_demo.ipynb]: Makes total sense, thank you. Added ORBAX_ITEMS_PATH to verify the full 0/items checkpoint path before proceeding.

File: src/maxtext/examples/gemma4_native_lora_demo.ipynb (hunk 3 above)
Comment (RexBearIU) [reply src/maxtext/examples/gemma4_native_lora_demo.ipynb]: Done, aligned the checkpoint check here too. Thanks again for the review.

File: src/maxtext/examples/native_lora_demo.ipynb
Comment (SurbhiJainUSC) [inline src/maxtext/examples/native_lora_demo.ipynb:1]: Since both `gemma4_native_lora_demo` and `qwen3_native_lora_demo` notebooks perform the exact same workflows, you could consolidate them into a single `native_lora_demo.ipynb` notebook. You can provide a configuration cell at the top where the user selects their preferred model. This would eliminate ~400 lines of duplicate code and make it much easier to maintain the notebooks.

File: docs/tutorials/posttraining/native_lora.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+## 🚀 Quick Experimentation with Notebooks
+
+For interactive playground setups on Google Colab or local JupyterLab, we provide fully detailed demo notebooks:
+
+- **Qwen3 Native LoRA Demo**: [qwen3_native_lora_demo.ipynb](https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/examples/qwen3_native_lora_demo.ipynb)
+- **Gemma4 Native LoRA Demo**: [gemma4_native_lora_demo.ipynb](https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/examples/gemma4_native_lora_demo.ipynb)
+
+______________________________________________________________________
+
```
Comment (SurbhiJainUSC) [inline docs/tutorials/posttraining/native_lora.md:47]: Also include installation step: https://maxtext.readthedocs.io/en/latest/tutorials/posttraining/sft.html#install-maxtext-and-post-training-dependencies

File: docs/tutorials/posttraining/native_lora.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+```bash
+hf auth login
+```
+
+Set the following environment variables before running LoRA Fine-tuning.
+
+```sh
+# -- Model configuration --
+export MODEL_NAME=<MODEL_NAME> # e.g., 'qwen3-0.6b' or 'gemma4-e2b'
```
Comment (SurbhiJainUSC) [inline docs/tutorials/posttraining/native_lora.md]: nit: `export MODEL=<MODEL_NAME>` for consistency with other docs.

File: docs/tutorials/posttraining/native_lora.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+Execute the following command to begin LoRA fine-tuning on a Hugging Face dataset (e.g. GSM8K) using the native SFT entrypoint `train_sft_native.py`:
+
+```sh
+python3 -m maxtext.trainers.post_train.sft.train_sft_native \
+    src/maxtext/configs/post_train/sft.yml \
+    run_name="${RUN_NAME?}" \
+    base_output_directory="${BASE_OUTPUT_DIRECTORY?}" \
+    model_name="${MODEL_NAME?}" \
+    load_parameters_path="${MAXTEXT_CKPT_PATH?}" \
+    tokenizer_path="${TOKENIZER_PATH?}" \
```
Comment (SurbhiJainUSC) [inline docs/tutorials/posttraining/native_lora.md]: Is there a need to set `TOKENIZER_PATH`? I think this can be derived automatically. Similarly `src/maxtext/configs/post_train/sft.yml`?

File: src/maxtext/examples/native_lora_demo.ipynb
Comment (RexBearIU) [reply src/maxtext/examples/native_lora_demo.ipynb:1]: Great idea, thank you. Consolidated both notebooks into a single native lora demo notebook with a model selection cell at the top for Gemma4 and Qwen3, removing duplicate code.

File: docs/tutorials/posttraining/native_lora.md (hunk 4 above)
Comment (RexBearIU) [reply docs/tutorials/posttraining/native_lora.md:47]: Added the installation section with a link to the official post training setup guide, thank you.

File: docs/tutorials/posttraining/native_lora.md (hunk 5 above)
Comment (RexBearIU) [reply docs/tutorials/posttraining/native_lora.md]: Updated to export MODEL and aligned the downstream training command for consistency, thank you.

File: docs/tutorials/posttraining/native_lora.md (hunk 6 above)
Comment (RexBearIU) [reply docs/tutorials/posttraining/native_lora.md]: Good point, thank you. Removed TOKENIZER_PATH and the explicit config YAML path so both are cleanly resolved by default.

File: docs/tutorials/posttraining/native_lora.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+______________________________________________________________________
+
+## Run Native Pre-training with QLoRA (8-bit Quantization)
+
+To run a standard native pre-training loop with memory-efficient 8-bit quantized weights, execute:
+
+```sh
+python3 -m maxtext.trainers.pre_train.train \
+    run_name="native_qlora_pretrain_demo" \
+    model_name="gemma4-e2b" \
```
Comment (SurbhiJainUSC) [inline docs/tutorials/posttraining/native_lora.md]: Should be `${MODEL?}`?

**Review by SurbhiJainUSC — APPROVED** [review]

File: docs/tutorials/posttraining/native_lora.md (hunk 7 above)
Comment (RexBearIU) [reply docs/tutorials/posttraining/native_lora.md]: Good catch, thank you. Updated model_name to use ${MODEL?} for consistency with the SFT command above.
